src/coordinator/storage/block.go

```diff
@@ -1,58 +1,157 @@
+		return multiSeriesBlock{}, err
+	}
+
+	meta := block.Metadata{
+		Bounds: block.Bounds{
+			Start:    query.Start,
+			End:      query.End,
+			StepSize: resolution,
+		},
+	}
+	return multiSeriesBlock{seriesList: seriesList, meta: meta}, nil
+}
+
+func (m multiSeriesBlock) Meta() block.Metadata {
+	return m.meta
+}
+
+func (m multiSeriesBlock) StepIter() block.StepIter {
+	return &multiSeriesBlockStepIter{block: m}
+}
+
+func (m multiSeriesBlock) SeriesIter() block.SeriesIter {
+	return &multiSeriesBlockSeriesIter{block: m}
 }
 
-// SeriesMeta is metadata data for the series
-type SeriesMeta struct {
-	Tags models.Tags
+func (m multiSeriesBlock) SeriesMeta() []block.SeriesMeta {
+	metas := make([]block.SeriesMeta, len(m.seriesList))
+	for i, s := range m.seriesList {
+		metas[i].Tags = s.Tags
+	}
+
+	return metas
+}
+
+func (m multiSeriesBlock) Close() error {
+	return nil
+}
+
+type multiSeriesBlockStepIter struct {
+	block multiSeriesBlock
+	index int
+}
+
+func (m *multiSeriesBlockStepIter) Next() bool {
+	if len(m.block.seriesList) == 0 {
+		return false
+	}
+
+	return m.index < m.block.seriesList[0].Values().Len()
+}
+
+func (m *multiSeriesBlockStepIter) Current() block.Step {
+	t := m.block.meta.Bounds.Start.Add(time.Duration(m.index) * m.block.meta.Bounds.StepSize)
+	values := make([]float64, len(m.block.seriesList))
+	for i, s := range m.block.seriesList {
+		values[i] = s.Values().ValueAt(i)
+	}
+
+	m.index++
+	return colStep{
+		t:      t,
+		values: values,
+	}
+}
+
+func (m *multiSeriesBlockStepIter) Len() int {
+	if len(m.block.seriesList) == 0 {
+		return 0
+	}
+
+	return m.block.seriesList[0].Values().Len()
 }
 
-// StepMeta is metadata data for a single time step
-type StepMeta struct {
+type colStep struct {
+	t      time.Time
+	values []float64
 }
 
-// Bounds are the time bounds
-// nolint: structcheck, megacheck, unused
-type Bounds struct {
-	start    time.Time
-	end      time.Time
-	stepSize time.Duration
+func (c colStep) Time() time.Time {
+	return c.t
 }
 
-// SeriesIter iterates through a CompressedSeriesIterator horizontally
-type SeriesIter interface {
-	Next() bool
-	Current() ts.Series
+func (c colStep) Values() []float64 {
+	return c.values
 }
 
-// StepIter iterates through a CompressedStepIterator vertically
-type StepIter interface {
-	Next() bool
-	Current() Step
+type multiSeriesBlockSeriesIter struct {
+	block multiSeriesBlock
+	index int
 }
 
-// Step can optionally implement iterator interface
-type Step interface {
-	Time() time.Time
-	Values() []float64
+func (m *multiSeriesBlockSeriesIter) Next() bool {
+	return m.index < len(m.block.seriesList)
 }
 
-// BlockMetadata is metadata for a block
-type BlockMetadata struct {
-	Bounds Bounds
-	Tags   models.Tags // Common tags across different series
+func (m *multiSeriesBlockSeriesIter) Current() block.Series {
+	s := m.block.seriesList[m.index]
+	values := make([]float64, s.Len())
+	for i := 0; i < s.Len(); i++ {
+		values[i] = s.Values().ValueAt(i)
```

Comment:
Any way we can memcopy this rather than copying one by one?

Reply:
that's gonna be hard since we hide Values() behind an interface instead of exposing the underlying slice. Will revisit later if this becomes a bottleneck